#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <mainwidget.h>
#include <QFileDialog>
#include<QMessageBox>
#include<QFile>
#include<QTime>
#include<QDateTime>
#include <QTableWidget>
#include<QKeyEvent>
#include<QChart>
#include<QChartView>
#include<QBarSet>
#include<QBarSeries>
#include<QBarCategoryAxis>
#include<QtDebug>
#include<QValueAxis>
#include<QLineSeries>
#include<QPieSeries>
#include <QtWidgets/QApplication>
#include <QtWidgets/QMainWindow>
#include "customslice.h"
#include <fstream>
#include "pentool.h"
#include "brushtool.h"
#include <QtWidgets/QLineEdit>
#include <QtWidgets/QGroupBox>
#include <QtWidgets/QDoubleSpinBox>
#include <QtWidgets/QFormLayout>
#include <QtWidgets/QFontDialog>
#include <QtWidgets/QPushButton>
#include <QtWidgets/QComboBox>
#include <QtWidgets/QCheckBox>
#include <iostream>
#include <string>
MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);
  \
 timer=new QTimer(this);
 connect(timer,SIGNAL(timeout()),SLOT(myfunction()));
 timer->start(100);
ui->tableWidget->setColumnCount(10);
ui->tableWidget->setRowCount(6100);
connect(ui->tableWidget,SIGNAL(cellEntered()),this,SLOT(insertRow()));

QBarSet *set0 = new QBarSet("Bob");


*set0 << 30 << 40 << 10 << 20 << 10 << 60;
set0->append(70);


QBarSeries *series = new QBarSeries();
series->append(set0);


QChart *chart = new QChart();
chart->addSeries(series);
chart->setTitle("BarChart Example In Qt5 C++ ");
chart->setAnimationOptions(QChart::SeriesAnimations);


QStringList categories;
categories << "Jan" << "Feb" << "Mar" << "Apr" << "May" << "Jun";
QBarCategoryAxis *axis = new QBarCategoryAxis();
axis->append(categories);
chart->createDefaultAxes();



QChartView *chartView = new QChartView(chart);
 chartView->resize(400,500);
chartView->setParent(ui->gridFrame);

}

MainWindow::~MainWindow()
{
    delete ui;
}


void MainWindow::on_actionOpen_triggered()
{

    QString file_name=QFileDialog::getOpenFileName(this,"open a file","C://");
    file_path=file_name.toStdString();
    std::string filename=file_name.toStdString();
    ui->plainTextEdit_3->setPlainText("File Name :-"+file_name);
    if(file_path.substr(file_path.find_last_of(".")+1)=="csv")
    Dataset *dataset= readCSV(filename);
}
Dataset* MainWindow::readCSV(std::string fileName)
{
    on_pushButton_4_clicked();
    // create dataset
    Dataset *dataset = new Dataset();

    // message window
    MessageWindow *messageWindow = new MessageWindow();

    // open the file
    std::ifstream input(fileName, std::ifstream::in);
    //input.clear();

    if (!input)
    {
        // error
        messageWindow->printError("CSV Reader",
                                  "Cannot open file",
                                  fileName);
        return nullptr;
    }


    // temporary container for strings
    std::string tmpStr = "";

    // read the file structure: comments, header and number of lines
    std::string headerStr = "";
    unsigned int nbLines = 0;

    while (getline(input, tmpStr))
    {
        // skip comments and empty lines
        if ( (tmpStr[0] != '#') && (!tmpStr.empty()) )
        {
            if (headerStr == "")
            {
                // header: supposed to be the first line after comments
                headerStr = tmpStr;
            }
            else
            {
                // new line of data
                nbLines += 1;
            }
        }
    }

    // the end of the file has been reached -> remove eof flag
    input.clear();

    // process header to get fields names
    std::vector<std::string> initHeaderFields = split(headerStr, ',');
    std::vector<std::string> headerFields = renameDuplicates(initHeaderFields);
    std::cout << "'" << headerStr << "'" << std::endl;
    for (unsigned int i=0; i<headerFields.size(); ++i)
    {
        std::cout << headerFields[i] << "   " << i << std::endl;
        headerFields[i] = strip(removeQuotes(headerFields[i]));
    }

    // create fields and dataset
    unsigned int nbFields = 0;
    dataset->setSize(nbLines);

    for (std::string str : headerFields)
    {
        Field newField;
        newField.setSize(nbLines);
        newField.setName(str);

        dataset->addField(newField);
        nbFields ++;
    }
    vector<string> *z[2];

    // go back to header and skip header
    input.clear();
    input.seekg(input.beg);
    input >> tmpStr;

    // read the data part line by line
    std::vector<std::string> splitStr;
    std::string fieldName;
    unsigned int lineNumber = 0;
    double value;
    Field tmpField;
int j=0;
    while (getline(input, tmpStr))
    {
        // skip comments and empty lines
        if ( (tmpStr[0] != '#') && (!tmpStr.empty()) )
        {
            // split the line and store pieces in the corresponding fields
            splitStr = split(tmpStr, ',');

            if (splitStr.size() != nbFields)
            {
                messageWindow->printError("CSV Reader",
                                          "Wrong number of columns");
                return nullptr;
            }
            if(splitStr.size()>m){
                ui->tableWidget->setColumnCount(splitStr.size());
                m=splitStr.size();
            }
            for (unsigned int i=0; i<splitStr.size(); ++i)
            {

                try
                {
                    value = std::stof(splitStr[i]);
                   // z[i]->push_back(to_string(value));
                    QTableWidgetItem *item = ui->tableWidget->item(j, i);
                    item = new QTableWidgetItem();
                    item->setText(QString::fromStdString(to_string(value)));
                    ui->tableWidget->setItem(j,i,item);


                }
                catch (invalid_argument)
                {
                    messageWindow->printError("CSV Reader","Cannot convert line data",splitStr[i]);
                    return nullptr;
                }
                dataset->setValue(headerFields[i], value, lineNumber);

            }
              cout<<j<<" "<<endl;;
            // increment line number
            lineNumber ++;
             j++;
        }

    }


    // set title
    std::string title;
    title = getBaseName(fileName);
    dataset->setTitle(title);
    cout<<j;

    // output
    return dataset;
}
void MainWindow::on_pushButton_clicked()
{

    QFile file(QString::fromStdString(file_path));
    if(!file.open(QFile::ReadOnly | QFile::Text)){
       QMessageBox::warning(this,"title","file not opened");

    }
    QTextStream in(&file);
    QString text=in.readAll();
    ui->plainTextEdit->setPlainText(text);
    file.close();
}
void MainWindow::myfunction(){
    QTime time=QTime::currentTime();
    QString time_text=time.toString("hh:mm:ss");
    ui->label_2->setText(time_text);
}

void MainWindow::on_actionSave_triggered()
{
    QFile file(QString::fromStdString(file_path));
   file.open(QFile::WriteOnly);
    QTextStream out(&file);
    QString text=ui->plainTextEdit->toPlainText();
   out<<text;
   file.flush();
    file.close();
}


void MainWindow::on_tabWidget_tabCloseRequested(int index)
{
    ui->tabWidget->removeTab(index);
}

void MainWindow::addRowIfLast()
{
    if(ui->tableWidget->currentRow() == ui->tableWidget->rowCount()-1) {
        ui->tableWidget->insertRow(ui->tableWidget->rowCount()+1);
        ui->tableWidget->setCurrentCell(ui->tableWidget->currentRow()+1, ui->tableWidget->currentColumn());

    } else
        ui->tableWidget->setCurrentCell(ui->tableWidget->currentRow()+1, ui->tableWidget->currentColumn());
}

void MainWindow::on_actionNew_triggered()
{
    ui->tabWidget->addTab(new QWidget(),QString("tab %0").arg(ui->tabWidget->count()+1));
    ui->tabWidget->addTab(ui->tab_2,QString("tab %0").arg(ui->tabWidget->count()+1));
}


void MainWindow::on_tableWidget_entered(const QModelIndex &index)
{

}


void MainWindow::on_tableWidget_cellEntered(int row, int column)
{
    ui->tableWidget->insertRow ( ui->tableWidget->rowCount());
}


void MainWindow::on_pushButton_3_clicked()

{

    int k=ui->spinBox->value()-1;
    int l=ui->spinBox->value()-1;
    QVector<QString> a;
    a.resize(ui->tableWidget->rowCount());
    QVector<double> b;
    QVector<double> d;
    d.resize(ui->tableWidget->rowCount());
    int c=0;
    b.resize(ui->tableWidget->rowCount());
    if(k==-1){
        for(int i=0;i<ui->tableWidget->rowCount();i++){
         a[i]="1";
         d[i]=1;
        }
    }
        if(l==-1){
            for(int i=0;i<ui->tableWidget->rowCount();i++){
             b[i]=1;
            }
        }

    for(int i=0;i<ui->tableWidget->rowCount();i++){
        if(k!=-1){
       if(ui->tableWidget->item(i,k)){
     a[i]=ui->tableWidget->item(i,k)->text();
     qDebug() <<a[i];
       }

       else
           a[i]=" ";
        }
  if(l!=-1){
   if(ui->tableWidget->item(i,l)){
       if(ui->tableWidget->item(i,l)->text()!=" "){
     QVariant val=ui->tableWidget->item(i,l)->text();
     b[i]=val.toDouble();

     if(c<b[i]){
         c=b[i];
     }
     qDebug()<<b[i];
       }
       else{
           b[i]=0;
           qDebug() <<b[i];
       }
   }
   else{
       b[i]=0;
   qDebug() <<b[i];
    }
}
  if(k!=-1){
   if(ui->tableWidget->item(i,k)){
       if(ui->tableWidget->item(i,k)->text()!=" "){
     QVariant val1=ui->tableWidget->item(i,k)->text();
     d[i]=val1.toDouble();
     qDebug() <<d[i];
       }
       else{
           d[i]=0;
       }
   }
   else{
       d[i]=0;
    }
}
    }
    QLineSeries *seriesl = new QLineSeries();
    for(int i=0;i<ui->tableWidget->rowCount();i++){
        seriesl->append(d[i],b[i]);
    }


     QBarSet *set1 = new QBarSet("A");
     QStringList categories1;
     for(int i=0;i<ui->tableWidget->rowCount();i++){
         qDebug() <<b[i];
     set1->append(b[i]);
     categories1.append(a[i]);
}

     QBarSeries *series1 = new QBarSeries();
     series1->append(set1);


     QChart *chart1 = new QChart();
     QChart *chart2 = new QChart();
     chart2->addSeries(seriesl);
     chart2->setTitle("BarChart Example In Qt5 C++ ");
     chart2->setAnimationOptions(QChart::SeriesAnimations);
     chart1->addSeries(series1);
     chart1->setTitle("BarChart Example In Qt5 C++ ");
     chart1->setAnimationOptions(QChart::SeriesAnimations);



     QBarCategoryAxis *axis1 = new QBarCategoryAxis();
     axis1->append(categories1);
     chart1->addAxis(axis1, Qt::AlignBottom);
     series1->attachAxis(axis1);
     QValueAxis *axisY = new QValueAxis();
     axisY->setRange(0,c);
     chart1->addAxis(axisY, Qt::AlignLeft);
     series1->attachAxis(axisY);
     chart1->legend()->setVisible(true);
     chart1->legend()->setAlignment(Qt::AlignTop);
     chart2->createDefaultAxes();
     QChartView *chartView1 = new QChartView(chart1);
     QChartView *chartView2 = new QChartView(chart2);
     ui->tabWidget->addTab(ui->tab,QString("tab %0").arg(ui->tabWidget->count()+1));
     chartView1->resize(400,500);
     chartView1->setParent(ui->gridFrame);
     chartView2->resize(400,500);
     chartView2->setParent(ui->gridFrame1);
     qDebug() <<ui->gridFrame->size();
     chartView1->setRenderHint(QPainter::Antialiasing);

     QPieSeries *series2 = new QPieSeries();
     for(int i=0;i<ui->tableWidget->rowCount();i++){
     series2->append(a[i], b[i]);
     }
     QChart *chart3= new QChart();
     chart3->addSeries(series2);
     chart3->setTitle("Simple piechart example");
     chart3->legend()->hide();

     QChartView *chartView3 = new QChartView(chart3);
     chartView3->resize(400,500);
     chartView3->setParent(ui->gridFrame2);
     chartView3->setRenderHint(QPainter::Antialiasing);

}



void MainWindow::on_actionLine_triggered()
{

    QVector<QString> a;
    a.resize(ui->tableWidget->rowCount());
    QVector<double> b;
    QVector<double> d;
    d.resize(ui->tableWidget->rowCount());
    int c=0;
    b.resize(ui->tableWidget->rowCount());
    for(int i=0;i<ui->tableWidget->rowCount();i++){
       if(ui->tableWidget->item(i,0)){
     a[i]=ui->tableWidget->item(i,0)->text();
     qDebug() <<a[i];
       }
       else
           a[i]=" ";

   if(ui->tableWidget->item(i,1)){
       if(ui->tableWidget->item(i,1)->text()!=" "){
     QVariant val=ui->tableWidget->item(i,1)->text();
     b[i]=val.toDouble();

     if(c<b[i]){
         c=b[i];
     }
     qDebug()<<b[i];
       }
       else{
           b[i]=0;
           qDebug() <<b[i];
       }
   }
   else{
       b[i]=0;
   qDebug() <<b[i];
    }

   if(ui->tableWidget->item(i,0)){
       if(ui->tableWidget->item(i,0)->text()!=" "){
     QVariant val1=ui->tableWidget->item(i,0)->text();
     d[i]=val1.toDouble();
     qDebug() <<d[i];
       }
       else{
           d[i]=0;
       }
   }
   else{
       d[i]=0;
    }

    }
    QLineSeries *seriesl = new QLineSeries();
    for(int i=0;i<ui->tableWidget->rowCount();i++){
        seriesl->append(d[i],b[i]);
    }
     QChart *chart2 = new QChart();
     chart2->addSeries(seriesl);
     chart2->setTitle("BarChart Example In Qt5 C++ ");
     chart2->setAnimationOptions(QChart::SeriesAnimations);
     chart2->createDefaultAxes();
     QChartView *chartView2 = new QChartView(chart2);
     QWidget *newTab = new QWidget(ui->tabWidget);
     ui->tabWidget->addTab(newTab,QString("tab %0").arg(ui->tabWidget->count()+1));
     chartView2->resize(1000,500);
     chartView2->setParent(newTab);
     //qDebug() <<ui->gridFrame->size();
     //QMainWindow window;
     //window.setCentralWidget(chartView2);
     //window.resize(400, 300);
     //window.show();
}


void MainWindow::on_actionBar_Chart_triggered()
{
    QVector<QString> a;
    a.resize(ui->tableWidget->rowCount());
    QVector<double> b;
    QVector<double> d;
    d.resize(ui->tableWidget->rowCount());
    int c=0;
    b.resize(ui->tableWidget->rowCount());
    for(int i=0;i<ui->tableWidget->rowCount();i++){
       if(ui->tableWidget->item(i,0)){
     a[i]=ui->tableWidget->item(i,0)->text();
     qDebug() <<a[i];
       }
       else
           a[i]=" ";

   if(ui->tableWidget->item(i,1)){
       if(ui->tableWidget->item(i,1)->text()!=" "){
     QVariant val=ui->tableWidget->item(i,1)->text();
     b[i]=val.toDouble();

     if(c<b[i]){
         c=b[i];
     }
     qDebug()<<b[i];
       }
       else{
           b[i]=0;
           qDebug() <<b[i];
       }
   }
   else{
       b[i]=0;
   qDebug() <<b[i];
    }

   if(ui->tableWidget->item(i,0)){
       if(ui->tableWidget->item(i,0)->text()!=" "){
     QVariant val1=ui->tableWidget->item(i,0)->text();
     d[i]=val1.toDouble();
     qDebug() <<d[i];
       }
       else{
           d[i]=0;
       }
   }
   else{
       d[i]=0;
    }

    }
     QBarSet *set1 = new QBarSet("A");
     QStringList categories1;
     for(int i=0;i<ui->tableWidget->rowCount();i++){
         qDebug() <<b[i];
     set1->append(b[i]);
     categories1.append(a[i]);
}

     QBarSeries *series1 = new QBarSeries();
     series1->append(set1);
     QChart *chart1 = new QChart();
     chart1->addSeries(series1);
     chart1->setTitle("BarChart Example In Qt5 C++ ");
     chart1->setAnimationOptions(QChart::SeriesAnimations);
     QBarCategoryAxis *axis1 = new QBarCategoryAxis();
     axis1->append(categories1);
     chart1->addAxis(axis1, Qt::AlignBottom);
     series1->attachAxis(axis1);
     QValueAxis *axisY = new QValueAxis();
     axisY->setRange(0,c);
     chart1->addAxis(axisY, Qt::AlignLeft);
     series1->attachAxis(axisY);
     chart1->legend()->setVisible(true);
     chart1->legend()->setAlignment(Qt::AlignTop);

     QChartView *chartView1 = new QChartView(chart1);

     qDebug() <<ui->gridFrame->size();
     chartView1->setRenderHint(QPainter::Antialiasing);
     QWidget *newTab = new QWidget(ui->tabWidget);
     ui->tabWidget->addTab(newTab,QString("tab %0").arg(ui->tabWidget->count()+1));
     chartView1->resize(1000,500);
     chartView1->setParent(newTab);

}


void MainWindow::on_actionPie_Chart_triggered()
{

    QVector<QString> a;
    a.resize(ui->tableWidget->rowCount());
    QVector<double> b;
    QVector<double> d;
    d.resize(ui->tableWidget->rowCount());
    int c=0;
    b.resize(ui->tableWidget->rowCount());
    for(int i=0;i<ui->tableWidget->rowCount();i++){
       if(ui->tableWidget->item(i,0)){
     a[i]=ui->tableWidget->item(i,0)->text();
     qDebug() <<a[i];
       }
       else
           a[i]=" ";

   if(ui->tableWidget->item(i,1)){
       if(ui->tableWidget->item(i,1)->text()!=" "){
     QVariant val=ui->tableWidget->item(i,1)->text();
     b[i]=val.toDouble();

     if(c<b[i]){
         c=b[i];
     }
     qDebug()<<b[i];
       }
       else{
           b[i]=0;
           qDebug() <<b[i];
       }
   }
   else{
       b[i]=0;
   qDebug() <<b[i];
    }

   if(ui->tableWidget->item(i,0)){
       if(ui->tableWidget->item(i,0)->text()!=" "){
     QVariant val1=ui->tableWidget->item(i,0)->text();
     d[i]=val1.toDouble();
     qDebug() <<d[i];
       }
       else{
           d[i]=0;
       }
   }
   else{
       d[i]=0;
    }

    }
    QPieSeries *series2 = new QPieSeries();
    for(int i=0;i<ui->tableWidget->rowCount();i++){
    series2->append(a[i], b[i]);
    }
    QChart *chart3= new QChart();
    chart3->addSeries(series2);
    chart3->setTitle("Simple piechart example");
    chart3->legend()->hide();

    QChartView *chartView3 = new QChartView(chart3);
    QWidget *newTab = new QWidget(ui->tabWidget);
    ui->tabWidget->addTab(newTab,QString("tab %0").arg(ui->tabWidget->count()+1));
    chartView3->resize(1000,500);
    chartView3->setParent(newTab);
    chartView3->setRenderHint(QPainter::Antialiasing);

   // QWidget *newTab = new QWidget(ui->tabWidget);
    //ui->tabWidget->addTab(newTab,QString("tab %0").arg(ui->tabWidget->count()+1));
     //connect(newTab,this)
    //MainWidget(this);
}

void MainWindow::on_pushButton_4_clicked()
{
    for(int i=0;i<ui->tableWidget->rowCount();i++){
        for(int j=0;j<ui->tableWidget->columnCount();j++){
            QTableWidgetItem *item = ui->tableWidget->item(i, j);
            item = new QTableWidgetItem();
            item->setText(QString::fromStdString(" "));
            ui->tableWidget->setItem(i,j,item);
        }
    }
}


void MainWindow::on_pushButton_2_clicked()
{
    QFile file(QString::fromStdString(file_path));
   file.open(QFile::WriteOnly);
    QTextStream out(&file);
    QString text=ui->plainTextEdit->toPlainText();
   out<<text;
   file.flush();
    file.close();
    Dataset *dataset= readCSV(file_path);
}

